postcode_lookup/template_sorter.py

```diff
@@ -258,7 +258,7 @@ def __init__(
             self.api_response, "registration", None
         )
         for i, date in enumerate(self.api_response.dates):
-            if parse(date.date) < datetime.datetime.today():
+            if parse(date.date).date() < datetime.datetime.today().date():
```

Comment:
I've looked at the test fail in the previous commit, but I don't understand how you got to this fix. Can you explain a little?

Reply:
Yeah sure. This bug was introduced in https://github.com/DemocracyClub/ec-postcode-lookup-pages/commit/638d5da93304988fe1dfb99ff341f2eb9bfb864e I assume this got added _after_ the GE otherwise this would have surely been noticed on polling day.

In this function, we set `self.dates` to an empty list:

https://github.com/DemocracyClub/ec-postcode-lookup-pages/blob/638d5da93304988fe1dfb99ff341f2eb9bfb864e/postcode_lookup/template_sorter.py#L253

Then we iterate over `self.api_response.dates` populating it.

If you load up this page, we always follow this branch at any time after midnight if the election is today because we're comparing 2 datetimes. So on the left, we parse a string like `2024-08-20` to a datetime. This always gives us `2024-08-20 00:00:00`. On the right side, `today()` basically gives us `now()` not midnight e.g: `2024-08-20 16:00:00` or whatever. So:

```
>>> import datetime
>>> from dateparser import parse
>>> print(parse('2024-08-20'))
2024-08-20 00:00:00
>>> print(datetime.datetime.today())
2024-08-20 16:34:15.939218
```

As soon as we're past midnight today, the datetime on the left is considered "less" than the datetime on the right. This means we never add anything to `self.dates` in this case. This then manifests as throwing a 500 due to an `IndexError: list index out of range` when we try to access `self.dates[0]` later on (Although the failing tests don't expose that - sorry I hadn't really clocked how useless the test output is there).

So what I'm doing here is just switching it so we are comparing 2 **dates** not 2 **datetimes**, which I think was the original intent of https://github.com/DemocracyClub/ec-postcode-lookup-pages/commit/638d5da93304988fe1dfb99ff341f2eb9bfb864e

Reply:
Sorry I've just realised "off by one" is not really a correct description of this error, which also probably didn't help.